GitHub API rate limit exceeded in `test_canned_staging_area`

https://github.com/DataBiosphere/azul/actions/runs/3521443550/jobs/5903293677  

~May be a dupe of~ Likely duplicated by https://github.com/DataBiosphere/azul/issues/4438

- [x] Security design review completed; the Resolution of this issue does **not** …
  - [x] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with 
  - [x] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [x] … affect the permissions of external users like access to
    - TDR snapshots
  - [x] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [x] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [x] … affect monitoring of the system
  - [x] … introduce a new software dependency like
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [x] … add an interface that exposes sensitive or confidential data at the security boundary
  - [x] … affect the encryption of data at rest
  - [x] … require persistence of sensitive or confidential data that might require encryption at rest
  - [x] … require unencrypted transmission of data within the security boundary
  - [x] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [x] Documentation on any unchecked boxes is provided in comments below

EDIT: This is in the wrong issue, see below.

It seems like the reference link in the description expired and the stack trace isn't accessible. But two recent builds have fail this test with different stack traces.

https://github.com/DataBiosphere/azul/actions/runs/4307226400/jobs/7512033686

```
======================================================================
ERROR: test_canned_staging_area (hca_metadata_api.test.TestAccessorApi)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/jsonschema/validators.py", line 909, in resolve_from_url
    document = self.store[url]
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/jsonschema/_utils.py", line 28, in __getitem__
    return self.store[self.normalize(uri)]
KeyError: 'https://schema.staging.data.humancellatlas.org/system/1.1.0/provenance'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/urllib3/connectionpool.py", line 703, in urlopen
    httplib_response = self._make_request(
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/urllib3/connectionpool.py", line 386, in _make_request
    self._validate_conn(conn)
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/urllib3/connectionpool.py", line 1042, in _validate_conn
    conn.connect()
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/urllib3/connection.py", line 414, in connect
    self.sock = ssl_wrap_socket(
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/urllib3/util/ssl_.py", line 449, in ssl_wrap_socket
    ssl_sock = _ssl_wrap_socket_impl(
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/urllib3/util/ssl_.py", line 493, in _ssl_wrap_socket_impl
    return ssl_context.wrap_socket(sock, server_hostname=server_hostname)
  File "/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/ssl.py", line 500, in wrap_socket
    return self.sslsocket_class._create(
  File "/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/ssl.py", line 1040, in _create
    self.do_handshake()
  File "/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/ssl.py", line 1309, in do_handshake
    self._sslobj.do_handshake()
ConnectionResetError: [Errno 104] Connection reset by peer

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/requests/adapters.py", line 489, in send
    resp = conn.urlopen(
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/urllib3/connectionpool.py", line 787, in urlopen
    retries = retries.increment(
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/urllib3/util/retry.py", line 550, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/urllib3/packages/six.py", line 769, in reraise
    raise value.with_traceback(tb)
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/urllib3/connectionpool.py", line 703, in urlopen
    httplib_response = self._make_request(
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/urllib3/connectionpool.py", line 386, in _make_request
    self._validate_conn(conn)
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/urllib3/connectionpool.py", line 1042, in _validate_conn
    conn.connect()
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/urllib3/connection.py", line 414, in connect
    self.sock = ssl_wrap_socket(
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/urllib3/util/ssl_.py", line 449, in ssl_wrap_socket
    ssl_sock = _ssl_wrap_socket_impl(
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/urllib3/util/ssl_.py", line 493, in _ssl_wrap_socket_impl
    return ssl_context.wrap_socket(sock, server_hostname=server_hostname)
  File "/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/ssl.py", line 500, in wrap_socket
    return self.sslsocket_class._create(
  File "/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/ssl.py", line 1040, in _create
    self.do_handshake()
  File "/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/ssl.py", line 1309, in do_handshake
    self._sslobj.do_handshake()
urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/jsonschema/validators.py", line 912, in resolve_from_url
    document = self.resolve_remote(url)
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/jsonschema/validators.py", line 1015, in resolve_remote
    result = requests.get(uri).json()
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/requests/api.py", line 73, in get
    return request("get", url, params=params, **kwargs)
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/requests/api.py", line 59, in request
    return session.request(method=method, url=url, **kwargs)
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/requests/sessions.py", line 587, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/requests/sessions.py", line 701, in send
    r = adapter.send(request, **kwargs)
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/requests/adapters.py", line 547, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/runner/work/azul/azul/test/hca_metadata_api/test.py", line 489, in test_canned_staging_area
    staging_area = factory.load_staging_area()
  File "/home/runner/work/azul/azul/src/humancellatlas/data/metadata/helpers/staging_area.py", line 279, in load_staging_area
    files = self._get_files(path=self.path + ('metadata', folder))
  File "/home/runner/work/azul/azul/src/humancellatlas/data/metadata/helpers/staging_area.py", line 316, in _get_files
    file = JsonFile.from_json(file_name, file_json)
  File "/home/runner/work/azul/azul/src/humancellatlas/data/metadata/helpers/staging_area.py", line 85, in from_json
    return MetadataFile(uuid=entity_id,
  File "<attrs generated init humancellatlas.data.metadata.helpers.staging_area.MetadataFile>", line 8, in __init__
    self.__attrs_post_init__()
  File "/home/runner/work/azul/azul/src/humancellatlas/data/metadata/helpers/staging_area.py", line 60, in __attrs_post_init__
    self._validator.validate_json(self.content, self.name)
  File "/home/runner/work/azul/azul/src/humancellatlas/data/metadata/helpers/schema_validation.py", line 34, in validate_json
    validate(file_json, schema, format_checker=FormatChecker())
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/jsonschema/validators.py", line 1119, in validate
    error = exceptions.best_match(validator.iter_errors(instance))
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/jsonschema/exceptions.py", line 384, in best_match
    best = next(errors, None)
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/jsonschema/validators.py", line 288, in iter_errors
    for error in errors:
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/jsonschema/_validators.py", line 332, in properties
    yield from validator.descend(
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/jsonschema/validators.py", line 305, in descend
    for error in self.evolve(schema=schema).iter_errors(instance):
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/jsonschema/validators.py", line 288, in iter_errors
    for error in errors:
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/jsonschema/_validators.py", line 294, in ref
    scope, resolved = validator.resolver.resolve(ref)
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/jsonschema/validators.py", line 898, in resolve
    return url, self._remote_cache(url)
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/jsonschema/validators.py", line 914, in resolve_from_url
    raise exceptions.RefResolutionError(exc)
jsonschema.exceptions.RefResolutionError: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))

----------------------------------------------------------------------
Ran 394 tests in 935.556s

FAILED (errors=1, skipped=4)
make: *** [Makefile:224: test] Error 1
Error: Process completed with exit code 2.
```

and https://github.com/DataBiosphere/azul/actions/runs/4307236496/jobs/7512056296

```
======================================================================
ERROR: test_canned_staging_area (hca_metadata_api.test.TestAccessorApi)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/requests/models.py", line 971, in json
    return complexjson.loads(self.text, **kwargs)
  File "/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/jsonschema/validators.py", line 912, in resolve_from_url
    document = self.resolve_remote(url)
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/jsonschema/validators.py", line 1015, in resolve_remote
    result = requests.get(uri).json()
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/requests/models.py", line 975, in json
    raise RequestsJSONDecodeError(e.msg, e.doc, e.pos)
requests.exceptions.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/runner/work/azul/azul/test/hca_metadata_api/test.py", line 489, in test_canned_staging_area
    staging_area = factory.load_staging_area()
  File "/home/runner/work/azul/azul/src/humancellatlas/data/metadata/helpers/staging_area.py", line 279, in load_staging_area
    files = self._get_files(path=self.path + ('metadata', folder))
  File "/home/runner/work/azul/azul/src/humancellatlas/data/metadata/helpers/staging_area.py", line 316, in _get_files
    file = JsonFile.from_json(file_name, file_json)
  File "/home/runner/work/azul/azul/src/humancellatlas/data/metadata/helpers/staging_area.py", line 85, in from_json
    return MetadataFile(uuid=entity_id,
  File "<attrs generated init humancellatlas.data.metadata.helpers.staging_area.MetadataFile>", line 8, in __init__
    self.__attrs_post_init__()
  File "/home/runner/work/azul/azul/src/humancellatlas/data/metadata/helpers/staging_area.py", line 60, in __attrs_post_init__
    self._validator.validate_json(self.content, self.name)
  File "/home/runner/work/azul/azul/src/humancellatlas/data/metadata/helpers/schema_validation.py", line 34, in validate_json
    validate(file_json, schema, format_checker=FormatChecker())
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/jsonschema/validators.py", line 1119, in validate
    error = exceptions.best_match(validator.iter_errors(instance))
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/jsonschema/exceptions.py", line 384, in best_match
    best = next(errors, None)
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/jsonschema/validators.py", line 288, in iter_errors
    for error in errors:
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/jsonschema/_validators.py", line 332, in properties
    yield from validator.descend(
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/jsonschema/validators.py", line 305, in descend
    for error in self.evolve(schema=schema).iter_errors(instance):
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/jsonschema/validators.py", line 288, in iter_errors
    for error in errors:
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/jsonschema/_validators.py", line 294, in ref
    scope, resolved = validator.resolver.resolve(ref)
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/jsonschema/validators.py", line 898, in resolve
    return url, self._remote_cache(url)
  File "/home/runner/work/azul/azul/.venv/lib/python3.9/site-packages/jsonschema/validators.py", line 914, in resolve_from_url
    raise exceptions.RefResolutionError(exc)
jsonschema.exceptions.RefResolutionError: Expecting value: line 1 column 1 (char 0)

----------------------------------------------------------------------
Ran 394 tests in 1346.841s

FAILED (errors=1, skipped=4)
make: *** [Makefile:224: test] Error 1
Error: Process completed with exit code 2.
```

--- a/src/azul/plugins/repository/canned/__init__.py
+++ b/src/azul/plugins/repository/canned/__init__.py
@@ -194,7 +194,6 @@ class Plugin(RepositoryPlugin[CannedBundle, SimpleSourceSpec, CannedSourceRef, C
                          file_uuid: str,
                          *,
                          file_version: Optional[str] = None,
-                         replica: Optional[str] = None,
                          ) -> Optional[str]:
         # Check all sources for the file. If a file_version was specified return
         # when we find a match, otherwise continue checking all sources and
@@ -243,8 +242,7 @@ class CannedFileDownload(RepositoryFileDownload):
                ) -> None:
         assert isinstance(plugin, Plugin)
         url = plugin._direct_file_url(file_uuid=self.file_uuid,
-                                      file_version=self.file_version,
-                                      replica=None)
+                                      file_version=self.file_version)
         self._location = url
 
     @property

--- a/src/azul/plugins/repository/canned/__init__.py
+++ b/src/azul/plugins/repository/canned/__init__.py
@@ -174,7 +174,7 @@ class Plugin(RepositoryPlugin[CannedBundle, SimpleSourceSpec, CannedSourceRef, C
         >>> source_url = 'https://github.com/USER/REPO/tree/REF/tests'
 
         >>> plugin._construct_file_url(source_url, 'foo.zip')
-        'https://github.com/USER/REPO/raw/REF/tests/foo.zip'
+        'https://github.com/USER/REPO/raw/REF/tests/data/foo.zip'
 
         >>> plugin._construct_file_url(source_url, '')
         Traceback (most recent call last):
@@ -184,6 +184,7 @@ class Plugin(RepositoryPlugin[CannedBundle, SimpleSourceSpec, CannedSourceRef, C
         url = furl(source_url)
         require(url.path.segments[2] == 'tree', source_url)
         url.path.segments[2] = 'raw'
+        url.path.segments.append('data')
         require(len(file_name) > 0, 'file_name cannot be empty')
         require(not file_name.endswith('/'), file_name)
         for segment in file_name.split('/'):

--- a/src/azul/plugins/repository/canned/__init__.py
+++ b/src/azul/plugins/repository/canned/__init__.py
@@ -2,13 +2,10 @@
 This repository plugin allows reading from a canned staging area like the one in
 the GitHub repo https://github.com/HumanCellAtlas/schema-test-data .
 
-NOTE: Use of this plugin requires a GitHub personal access token specified in an
-environment variable `GITHUB_TOKEN`. See:
-https://github.com/DataBiosphere/hca-metadata-api#github-credentials
-
-Due to this requirement, this plugin cannot be used to index data directly from
-the canned staging area, however it can be used with the `can_bundle.py` script
-to create a local canned bundle from files in the canned staging area.
+NOTE: This plugin's purpose is for testing and verification of a canned staging
+area, and should not be used to create catalogs on a deployment. It can however
+be used with the `can_bundle.py` script to create a local canned bundle from the
+files in the canned staging area.
 """
 from collections.abc import (
     Sequence,
@@ -18,6 +15,12 @@ from dataclasses import (
     dataclass,
 )
 import logging
+from pathlib import (
+    Path,
+)
+from tempfile import (
+    TemporaryDirectory,
+)
 import time
 from typing import (
     Optional,
@@ -68,7 +71,7 @@ from azul.uuids import (
     validate_uuid_prefix,
 )
 from humancellatlas.data.metadata.helpers.staging_area import (
-    GitHubStagingAreaFactory,
+    CannedStagingAreaFactory,
     StagingArea,
 )
 
@@ -120,12 +123,51 @@ class Plugin(RepositoryPlugin[CannedBundle, SimpleSourceSpec, CannedSourceRef, C
         ]
 
     def _lookup_source_id(self, spec: SimpleSourceSpec) -> str:
-        return spec.name
+        return str(spec)
+
+    def parse_github_url(self, url: furl) -> tuple[furl, Path, str]:
+        """
+        Parse a GitHub URL.
+
+        :param url: A GitHub URL of the format
+                    https://github.com/<OWNER>/<NAME>/tree/<REF>[/<PATH>]. Note
+                    that REF can be the name of a branch, the name of a tag, or
+                    a commit SHA. If REF contains special characters like `/`,
+                    '?` or `#` they must be URL-encoded. This is especially
+                    noteworthy for `/` in branch names.
+
+        :return: A tuple containing the URL of a GitHub repository, a relative
+                 path inside that repository, and a Git ref.
+
+        >>> plugin = Plugin(_sources=set())
+
+        >>> plugin.parse_github_url(furl('https://github.com/OWNER/NAME/tree/REF/tests'))
+        (furl('https://github.com/OWNER/NAME.git'), PosixPath('tests'), 'REF')
+        """
+        require(url.scheme == 'https', url)
+        require(url.host == 'github.com', url)
+        owner, name, slug, ref, *path = url.path.segments
+        require(slug == 'tree', url)
+        remote_url = furl(url.origin)
+        remote_url.path.add((owner, f'{name}.git'))
+        return remote_url, Path(*path), ref
 
     @lru_cache
-    def staging_area(self, source_spec: SimpleSourceSpec) -> StagingArea:
-        factory = GitHubStagingAreaFactory.from_url(source_spec.name)
-        return factory.load_staging_area()
+    def staging_area(self, url: str) -> StagingArea:
+        """
+        Process the contents of a staging area.
+
+        :param url: The URL of a staging area located in a GitHub repository.
+
+        :return: A StagingArea object containing the contents of the staging
+                 area's JSON files.
+        """
+        with TemporaryDirectory() as tmpdir:
+            remote_url, path, ref = self.parse_github_url(furl(url))
+            factory = CannedStagingAreaFactory.clone_remote(remote_url,
+                                                            Path(tmpdir),
+                                                            ref)
+            return factory.load_staging_area(path)
 
     def _assert_source(self, source: CannedSourceRef):
         assert source.spec in self.sources, (source, self.sources)
@@ -139,7 +181,8 @@ class Plugin(RepositoryPlugin[CannedBundle, SimpleSourceSpec, CannedSourceRef, C
         validate_uuid_prefix(prefix)
         log.info('Listing bundles with prefix %r in source %r.', prefix, source)
         bundle_fqids = []
-        for link in self.staging_area(source.spec).links.values():
+        staging_area = self.staging_area(source.spec.name)
+        for link in staging_area.links.values():
             if link.uuid.startswith(prefix):
                 bundle_fqids.append(CannedBundleFQID(source=source,
                                                      uuid=link.uuid,
@@ -151,8 +194,8 @@ class Plugin(RepositoryPlugin[CannedBundle, SimpleSourceSpec, CannedSourceRef, C
     def fetch_bundle(self, bundle_fqid: CannedBundleFQID) -> CannedBundle:
         self._assert_source(bundle_fqid.source)
         now = time.time()
-        staging_area = self.staging_area(bundle_fqid.source.spec)
-        version, manifest, metadata = staging_area.get_bundle(bundle_fqid.uuid)
+        staging_area = self.staging_area(bundle_fqid.source.spec.name)
+        version, manifest, metadata = staging_area.get_bundle_parts(bundle_fqid.uuid)
         if bundle_fqid.version is None:
             bundle_fqid = CannedBundleFQID(source=bundle_fqid.source,
                                            uuid=bundle_fqid.uuid,
@@ -168,55 +211,56 @@ class Plugin(RepositoryPlugin[CannedBundle, SimpleSourceSpec, CannedSourceRef, C
     def portal_db(self) -> Sequence[JSON]:
         return []
 
-    def _construct_file_url(self, source_url: str, file_name: str) -> str:
+    def _construct_file_url(self, url: furl, file_name: str) -> furl:
         """
         >>> plugin = Plugin(_sources=set())
-        >>> source_url = 'https://github.com/USER/REPO/tree/REF/tests'
+        >>> url = furl('https://github.com/OWNER/REPO/tree/REF/tests')
 
-        >>> plugin._construct_file_url(source_url, 'foo.zip')
-        'https://github.com/USER/REPO/raw/REF/tests/data/foo.zip'
+        >>> plugin._construct_file_url(url, 'foo.zip')
+        furl('https://github.com/OWNER/REPO/raw/REF/tests/data/foo.zip')
 
-        >>> plugin._construct_file_url(source_url, '')
+        >>> plugin._construct_file_url(url, '')
         Traceback (most recent call last):
         ...
         azul.RequirementError: file_name cannot be empty
         """
-        url = furl(source_url)
-        require(url.path.segments[2] == 'tree', source_url)
-        url.path.segments[2] = 'raw'
-        url.path.segments.append('data')
+        require(url.path.segments[2] == 'tree', str(url))
+        file_url = furl(url)
+        file_url.path.segments[2] = 'raw'
+        file_url.path.segments.append('data')
         require(len(file_name) > 0, 'file_name cannot be empty')
         require(not file_name.endswith('/'), file_name)
         for segment in file_name.split('/'):
-            url.path.segments.append(segment)
-        return str(url)
+            file_url.path.segments.append(segment)
+        return file_url
 
     def _direct_file_url(self,
                          file_uuid: str,
                          *,
                          file_version: Optional[str] = None,
-                         ) -> Optional[str]:
+                         ) -> Optional[furl]:
         # Check all sources for the file. If a file_version was specified return
         # when we find a match, otherwise continue checking all sources and
         # return the URL for the match with the latest (largest) version.
         found_version = None
         found_url = None
         for source_spec in self.sources:
-            staging_area = self.staging_area(source_spec)
+            staging_area = self.staging_area(source_spec.name)
             try:
                 descriptor = staging_area.descriptors[file_uuid]
             except KeyError:
                 continue
             else:
+                staging_area_url = furl(source_spec.name)
                 actual_file_version = descriptor.content['file_version']
                 if file_version:
                     if file_version == actual_file_version:
                         file_name = descriptor.content['file_name']
-                        return self._construct_file_url(source_spec.name, file_name)
+                        return self._construct_file_url(staging_area_url, file_name)
                 else:
                     if found_version is None or actual_file_version > found_version:
                         file_name = descriptor.content['file_name']
-                        found_url = self._construct_file_url(source_spec.name, file_name)
+                        found_url = self._construct_file_url(staging_area_url, file_name)
                         found_version = actual_file_version
         return found_url
 
@@ -234,7 +278,7 @@ class Plugin(RepositoryPlugin[CannedBundle, SimpleSourceSpec, CannedSourceRef, C
 
 
 class CannedFileDownload(RepositoryFileDownload):
-    _location: Optional[str] = None
+    _location: Optional[furl] = None
     _retry_after: Optional[int] = None
 
     def update(self,
@@ -248,7 +292,7 @@ class CannedFileDownload(RepositoryFileDownload):
 
     @property
     def location(self) -> Optional[str]:
-        return self._location
+        return None if self._location is None else str(self._location)
 
     @property
     def retry_after(self) -> Optional[int]:

--- a/src/humancellatlas/data/metadata/helpers/staging_area.py
+++ b/src/humancellatlas/data/metadata/helpers/staging_area.py
@@ -3,16 +3,16 @@ from collections import (
 )
 import json
 import logging
-import os
+from pathlib import (
+    Path,
+)
 from typing import (
     ClassVar,
-    List,
     Mapping,
-    Optional,
+    Self,
     Sequence,
     Set,
-    Tuple,
-    Union,
+    TypeVar,
 )
 from uuid import (
     UUID,
@@ -23,15 +23,11 @@ import attr
 from furl import (
     furl,
 )
-from github import (
-    Github,
-    UnknownObjectException,
-)
-from github.Repository import (
-    Repository,
-)
+
+import git
 
 from humancellatlas.data.metadata.api import (
+    Bundle,
     JSON,
 )
 from humancellatlas.data.metadata.helpers.exception import (
@@ -138,15 +134,17 @@ class StagingArea:
     metadata: Mapping[str, MetadataFile]  # Key is the entity ID
     descriptors: Mapping[str, DescriptorFile]  # Key is the entity ID
 
-    def get_bundle(self, subgraph_id: str) -> Tuple[str, List[JSON], JSON]:
+    def get_bundle(self, subgraph_id: str) -> Bundle:
         """
-        Return a tuple consisting of the version of the downloaded bundle, a
-        list of the manifest entries for all metadata files in the bundle, and a
-        dictionary mapping the file name of each metadata file in the bundle to
-        the JSON contents of that file.
+        Return a bundle from the staging area
         """
-        log.debug('Composing bundle %s', subgraph_id)
+        version, manifest, metadata = self.get_bundle_parts(subgraph_id)
+        return Bundle(subgraph_id, version, manifest, metadata)
 
+    def get_bundle_parts(self, subgraph_id: str) -> tuple[str, list[JSON], JSON]:
+        """
+        Return the components to create a bundle from the staging area
+        """
         links_file = self.links[subgraph_id]
         manifest = []
         metadata = {
@@ -178,15 +176,14 @@ class StagingArea:
                     manifest.append(file_manifest)
                 else:
                     pass
-        log.debug('Composed bundle with %i metadata files', len(metadata))
         return links_file.version, manifest, metadata
 
     def _entity_ids_by_type(self,
                             subgraph_id: str
-                            ) -> Mapping[str, Set[str]]:
+                            ) -> dict[str, Set[str]]:
         """
-        Parse the links in a subgraph and return a mapping of the entity types
-        (e.g. 'analysis_file', 'cell_suspension') to a set of entity IDs
+        Return a mapping of entity types (e.g. 'analysis_file',
+        'cell_suspension') to a set of entity IDs
         """
         links_file: LinksFile = self.links[subgraph_id]
         links_json = links_file.content
@@ -217,119 +214,113 @@ class StagingArea:
         return entity_ids
 
 
+JSON_FILE = TypeVar('JSON_FILE', bound=JsonFile)
+
+
 @attr.s(frozen=True, kw_only=True, auto_attribs=True)
-class GitHubStagingAreaFactory:
-    repo: Repository
-    ref: str
-    path: Tuple[str, ...]
+class CannedStagingAreaFactory:
+    #: Path to a local directory containing one or more staging areas
+    base_path: Path
 
     @classmethod
-    def create(cls,
-               owner: str,
-               name: str,
-               ref: str,
-               path: Optional[str] = None
-               ) -> 'GitHubStagingAreaFactory':
+    def clone_remote(cls, remote_url: furl, local_path: Path, ref: str) -> Self:
         """
-        :param owner: The owner of the GitHub repository
-        :param name: The name of the GitHub repository
-        :param ref: A branch name, tag, or commit SHA
-        :param path: The path inside the repository to the base of the staging area
-        """
-        token = os.environ['GITHUB_TOKEN']  # A GitHub personal access token
-        log.debug('Requesting GitHub repo %s', (owner, name, ref, path))
-        github_api = Github(token)
-        rate_limit = github_api.get_rate_limit()
-        log.debug('GitHub API rate limit limit=%i, remaining=%i, reset=%s',
-                  rate_limit.core.limit, rate_limit.core.remaining, rate_limit.core.reset)
-        repo = github_api.get_repo(f'{owner}/{name}')
-        path = () if path == '' else tuple(path.split('/'))
-        return cls(repo=repo, ref=ref, path=path)
+        Clone a remote Git repository and return a factory for staging areas
+        inside that clone.
 
-    @classmethod
-    def from_url(cls, url: str) -> 'GitHubStagingAreaFactory':
-        """
-        :param url: The URL of a staging area in a GitHub repository with syntax
-                    `https://github.com/<OWNER>/<NAME>/tree/<REF>[/<PATH>]`.
-                    Note that REF can be a branch, tag, or commit SHA. If REF
-                    contains special characters like `/`, '?` or `#` they must
-                    be URL-encoded. This is especially noteworthy for `/` since
-                    it's the only way to distinguish slashes in REF from those
-                    in PATH. The slashes in PATH must not be URL-encoded, while
-                    occurrences of `#` and `?` must.
-        """
-        parsed_url = furl(url)
-        require(parsed_url.scheme == 'https', url)
-        require(parsed_url.host == 'github.com', url)
-        require(len(parsed_url.path.segments) > 3, url)
-        require(parsed_url.path.segments[2] == 'tree', url)
-        owner, name = parsed_url.path.segments[0:2]
-        ref = parsed_url.path.segments[3]
-        path = '/'.join(parsed_url.path.segments[4:])
-        return cls.create(owner=owner, name=name, ref=ref, path=path)
+        :param remote_url: The URL of a remote Git repository containing one or
+                           more staging areas
 
-    def load_staging_area(self) -> StagingArea:
-        staging_area_folders = self._get_folders(path=self.path)
+        :param local_path: The path to an empty local directory where the
+                           repository will be cloned
+
+        :param ref: A Git ref (branch, tag, or commit SHA)
+        """
+        log.debug('Cloning %s into %s', remote_url, local_path)
+        repo = git.Repo.clone_from(str(remote_url), local_path)
+        log.debug('Checking out ref %s', ref)
+        repo.git.checkout(ref)
+        return cls(base_path=local_path)
+
+    def load_staging_area(self, path: Path) -> StagingArea:
+        """
+        Create and return a staging area object from the files in a local
+        staging area.
+
+        :param path: The relative path from `self.base_path` to a local staging
+                     area
+        """
+        path = self.base_path / path
+        staging_area_folders = {p.name for p in path.iterdir()}
         expected_folders = {'data', 'descriptors', 'links', 'metadata'}
-        require(set(staging_area_folders) == expected_folders,
-                f'{self.repo.full_name} {self.path} is not a valid staging area')
-        links = self._get_files(path=self.path + ('links',))
-        metadata = {}
-        for folder in self._get_folders(path=self.path + ('metadata',)):
-            files = self._get_files(path=self.path + ('metadata', folder))
-            metadata.update(files)
-        descriptors = {}
-        for folder in self._get_folders(path=self.path + ('descriptors',)):
-            files = self._get_files(path=self.path + ('descriptors', folder))
-            descriptors.update(files)
-        return StagingArea(links=links, metadata=metadata, descriptors=descriptors)
+        require(expected_folders == staging_area_folders,
+                'Invalid staging area', path)
+        return StagingArea(links=self._get_link_files(path),
+                           metadata=self._get_metadata_files(path),
+                           descriptors=self._get_descriptor_files(path))
 
-    def _get_folders(self, path: Tuple[str, ...]) -> List[str]:
-        folders = []
-        log.debug('Getting list of folders in %s', path)
-        try:
-            path_contents = self.repo.get_contents(path='/'.join(path), ref=self.ref)
-        except UnknownObjectException as e:
-            raise ValueError('Github path not found', self.repo.full_name, self.path) from e
-        for content in path_contents:
-            if content.type == 'dir':
-                folders.append(content.name)
-        return folders
+    def _get_link_files(self, path: Path) -> dict[str, LinksFile]:
+        """
+        Return a mapping of file ID to file content for all the link files in
+        the staging area.
+        """
+        return self._get_files(path=path / 'links', file_cls=LinksFile)
+
+    def _get_metadata_files(self, path: Path) -> dict[str, MetadataFile]:
+        """
+        Return a mapping of file ID to file content for all the metadata files
+        in the staging area.
+        """
+        files = {}
+        for sub_dir in (path / 'metadata').iterdir():
+            assert sub_dir.is_dir()
+            files.update(self._get_files(path=sub_dir, file_cls=MetadataFile))
+        return files
+
+    def _get_descriptor_files(self, path: Path) -> dict[str, DescriptorFile]:
+        """
+        Return a mapping of file ID to file content for all the descriptor files
+        in the staging area.
+        """
+        files = {}
+        for sub_dir in (path / 'descriptors').iterdir():
+            assert sub_dir.is_dir()
+            files.update(self._get_files(path=sub_dir, file_cls=DescriptorFile))
+        return files
 
     def _get_files(self,
-                   path: Tuple[str, ...],
-                   ) -> Mapping[str, Union[LinksFile, MetadataFile, DescriptorFile]]:
+                   path: Path,
+                   file_cls: type[JSON_FILE]
+                   ) -> dict[str, JSON_FILE]:
+        """
+        Return a mapping of file ID to file content for all the files found in
+        the directory at the given path.
+        """
         files = {}
-        log.debug('Getting contents of %s', path)
-        path_str = '/'.join(path)
+        log.debug('Reading files in %s', path)
+        for file in path.iterdir():
+            assert file.is_file()
+            with open(file, 'r') as f:
+                content = json.load(f)
+            file_name = file.name
+            json_file = JsonFile.from_json(file_name, content)
+            require(isinstance(json_file, file_cls), json_file)
+            self._add_file(files, json_file)
+        return files
+
+    def _add_file(self, files: dict[str, JSON_FILE], file: JSON_FILE) -> None:
+        """
+        Add `file` to `files`. If a file with the same ID already exists in
+        `files`, the file with the most recent version will be kept.
+        """
         try:
-            path_contents = self.repo.get_contents(path=path_str, ref=self.ref)
-        except UnknownObjectException as e:
-            raise ValueError('Github path not found', self.repo.full_name, path_str) from e
-        for content in path_contents:
-            if content.type == 'dir':
-                log.warning('Unexpected folder %s found in %s/%s',
-                            content.name, self.repo.full_name, path_str)
-                continue
-            file_name = content.name
-            file_json = json.loads(content.decoded_content)
-            file = JsonFile.from_json(file_name, file_json)
-            if isinstance(file, LinksFile):
-                require(path[-1] == 'links', content.path)
-            elif isinstance(file, DescriptorFile):
-                require(path[-2] == 'descriptors', content.path)
-            else:
-                require(path[-2] == 'metadata', content.path)
-            try:
-                existing_version = files[file.uuid].version
-            except KeyError:
+            existing_version = files[file.uuid].version
+        except KeyError:
+            files[file.uuid] = file
+        else:
+            reject(file.version == existing_version, file)
+            if file.version > existing_version:
                 files[file.uuid] = file
             else:
-                # If multiple files with the same ID were found keep the
-                # one with the largest version.
-                reject(file.version == existing_version, file)
-                if file.version > existing_version:
-                    files[file.uuid] = file
-                else:
-                    pass
-        return files
+                log.debug('Discarding previous %s version of file %s',
+                          existing_version, file)

--- a/test/hca_metadata_api/test.py
+++ b/test/hca_metadata_api/test.py
@@ -12,6 +12,12 @@ from itertools import (
 import json
 import logging
 import os
+from pathlib import (
+    Path,
+)
+from tempfile import (
+    TemporaryDirectory,
+)
 from unittest import (
     skip,
 )
@@ -21,6 +27,9 @@ from uuid import (
 
 from atomicwrites import (
     atomic_write,
+)
+from furl import (
+    furl,
 )
 from more_itertools import (
     one,
@@ -53,7 +62,7 @@ from humancellatlas.data.metadata.helpers.json import (
     as_json,
 )
 from humancellatlas.data.metadata.helpers.staging_area import (
-    GitHubStagingAreaFactory,
+    CannedStagingAreaFactory,
 )
 
 
@@ -488,19 +497,19 @@ class TestAccessorApi(AzulUnitTestCase):
 
         return bundle
 
-    # FIXME: Re-enable this test
-    #        https://github.com/DataBiosphere/azul/issues/4757
-    @skip('https://github.com/DataBiosphere/azul/issues/4757')
     def test_canned_staging_area(self):
+        remote_url = furl('https://github.com/HumanCellAtlas/schema-test-data.git')
+        staging_area_path = Path('tests')
         ref = 'eb93f83b'
-        url = f'https://github.com/HumanCellAtlas/schema-test-data/tree/{ref}/tests'
-        factory = GitHubStagingAreaFactory.from_url(url)
-        staging_area = factory.load_staging_area()
+        with TemporaryDirectory() as tmpdir:
+            factory = CannedStagingAreaFactory.clone_remote(remote_url,
+                                                            Path(tmpdir),
+                                                            ref)
+            staging_area = factory.load_staging_area(staging_area_path)
         self.assertGreater(len(staging_area.links), 0)
         for link_id in staging_area.links:
             with self.subTest(link_id=link_id):
-                version, manifest, metadata_files = staging_area.get_bundle(link_id)
-                bundle = Bundle(link_id, version, manifest, metadata_files)
+                bundle = staging_area.get_bundle(link_id)
                 self.assertEqual(bundle.uuid, UUID(link_id))
                 project = bundle.projects[UUID('90bf705c-d891-5ce2-aa54-094488b445c6')]
                 self.assertEqual(project.estimated_cell_count, 10000)

--- a/test/test_doctests.py
+++ b/test/test_doctests.py
@@ -34,6 +34,7 @@ import azul.openapi.responses
 import azul.openapi.schema
 import azul.plugins.metadata.hca.indexer.transform
 import azul.plugins.metadata.hca.service.contributor_matrices
+import azul.plugins.repository.canned
 import azul.plugins.repository.tdr_hca
 import azul.service.drs_controller
 import azul.service.manifest_service
@@ -83,6 +84,7 @@ def load_tests(_loader, tests, _ignore):
         azul.openapi.responses,
         azul.openapi.schema,
         azul.plugins.metadata.hca.service.contributor_matrices,
+        azul.plugins.repository.canned,
         azul.plugins.repository.tdr_hca,
         azul.plugins.metadata.hca.indexer.transform,
         azul.service.drs_controller,

